flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/window/slicing/SliceAssigners.java

```diff
@@ -0,0 +1,472 @@
+            // null namespace means use heap data views, instead of state state views
+            callback.merge(null, toBeMerged);
+        }
+
+        @Override
+        public Optional<Long> nextTriggerWindow(long windowEnd, Supplier<Boolean> isWindowEmpty) {
+            if (isWindowEmpty.get()) {
+                return Optional.empty();
+            } else {
+                return Optional.of(windowEnd + sliceSize);
+            }
+        }
+    }
+
+    /** The {@link SliceAssigner} for cumulative windows. */
+    public static final class CumulativeSliceAssigner extends AbstractSliceAssigner
+            implements SliceSharedAssigner {
+        private static final long serialVersionUID = 1L;
+
+        /** Creates a new {@link CumulativeSliceAssigner} with a new specified offset. */
+        public CumulativeSliceAssigner withOffset(Duration offset) {
+            return new CumulativeSliceAssigner(rowtimeIndex, maxSize, step, offset.toMillis());
+        }
+
+        private final long maxSize;
+        private final long step;
+        private final long offset;
+        private final ReusableListIterable reuseList = new ReusableListIterable();
+
+        protected CumulativeSliceAssigner(int rowtimeIndex, long maxSize, long step, long offset) {
+            super(rowtimeIndex);
+            if (maxSize <= 0 || step <= 0) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Cumulative Window parameters must satisfy maxSize > 0 and step > 0, but got maxSize %dms and step %dms.",
+                                maxSize, step));
+            }
+            if (maxSize % step != 0) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Cumulative Window requires maxSize must be an integral multiple of step, but got maxSize %dms and step %dms.",
+                                maxSize, step));
+            }
+
+            this.maxSize = maxSize;
+            this.step = step;
+            this.offset = offset;
+        }
+
+        @Override
+        public long assignSliceEnd(long timestamp) {
+            long start = TimeWindow.getWindowStartWithOffset(timestamp, offset, step);
+            return start + step;
+        }
+
+        @Override
+        public long getWindowStart(long windowEnd) {
+            return TimeWindow.getWindowStartWithOffset(windowEnd - 1, offset, maxSize);
+        }
+
+        @Override
+        public Iterable<Long> expiredSlices(long windowEnd) {
+            long windowStart = getWindowStart(windowEnd);
+            long firstSliceEnd = windowStart + step;
+            long lastSliceEnd = windowStart + maxSize;
+            if (windowEnd == firstSliceEnd) {
+                // we reuse state in the first slice, skip cleanup for the first slice
+                return Collections.emptyList();
```

Comment:
Unfortunately, we can't. Because the `reuseList` is not only used to construct expried slices but also the merged slices. I think it's fine to think it as a resued wrapper of result.